src/sds.c

```diff
@@ -195,7 +195,7 @@ sds sdsdup(const sds s) {
 /* Free an sds string. No operation is performed if 's' is NULL. */
 void sdsfree(sds s) {
     if (s == NULL) return;
-    s_free((char *)s - sdsHdrSize(s[-1]));
+    s_free_with_size(sdsAllocPtr(s), sdsAllocSize(s));
```

Comment:
I feel this is ugly AND I feel this is not your fault.

The mix between requested length and allocated size is non intuitive and error prone.

I suggest:
1/  sdsalloc() will explicitly refuse to work for SDS_TYPE_5 by asserting 
`   case SDS_TYPE_5: assert(0, "Type 5 does not have alloc")`
which is fair as Type 5 does not have alloc value and any answer is wrong

2/ similarly sdsAllocSize function is simply and will fail spectacularly on Type 5
```
size_t sdsAllocSize(sds s) {
    return sdsHdrSize(s[-1] & SDS_TYPE_MASK) + sdsalloc(s) + 1;
}
 // how come we dont have sdsTYPE(s) macro?
// Why sdsHdrSize does not take SDS but take a type?
// Why dont we have sdsHdrSizeByType
// So many questions...
```
And thus, should never be called for TYPE_5

3/ If its interesting anywhere to anyone... add a function which explicitly fuse the meaning and with a very descriptive comment as to what it actually returns 
```
size_t sdsAllocSizeOrLen(sds s) {
   if (s[-1] & SDS_TYPE_MASK == SDS_TYPE_5) { 
       return sdsLen(s) + 1;
   }
   return sdsAllocSize(s);
}
```
I suspect this function is not really needed at all 

4/ And just here in `sdsFree` do the if else...
```
void sdsfree(sds s) {
    if (s == NULL) return; // By the way this is lazy we should not allow sdsfree-ing a null pointer... but thats a point for future PR as there is probably code that relay on this thing 
   if (flags & SDS_TYPE_MASK == SDS_TYPE_5) { 
      s_free((char *)s - 1);
   } else {
     s_free_with_size(sdsAllocPtr(s), sdsAllocSize(s[-1]));
  }
```

Reply:
Just peeked and following my suggestion will actually unnecessarily bloat this PR.

So I am OK with the PR as is... and its worth considering a future cleanup in a separate PR

Reply:
> I feel this is ugly AND I feel this is not your fault.
> 
> The mix between requested length and allocated size is non intuitive and error prone.

I understand the complexity part, but I think it is common to consider allocation size in order to avoid extra fragmentation. I don't think this change in the PR is ugly, it actually aligned the TYPE_5 logic with all other types. My only concern was that now we will use REAL malloc_size whenever we question TYPE_5 allocation size (in the CURRENT code I did not find any real concerns though).

> 
> I suggest: 1/ sdsalloc() will explicitly refuse to work for SDS_TYPE_5 by asserting ` case SDS_TYPE_5: assert(0, "Type 5 does not have alloc")` which is fair as Type 5 does not have alloc value and any answer is wrong
> 
> 2/ similarly sdsAllocSize function is simply and will fail spectacularly on Type 5
> 
> ```
> size_t sdsAllocSize(sds s) {
>     return sdsHdrSize(s[-1] & SDS_TYPE_MASK) + sdsalloc(s) + 1;
> }

IMO this will create more complexity and potential bugs which are not covered by tests. I agree asserting in order to prevent the use for TYPE_5 is a much more explicit way to prevent use of it for TYPE_5 but the dynamic nature of sds sizes will make it hard to maintain.  

>  // how come we dont have sdsTYPE(s) macro?

+1. we should add that. 

> // Why sdsHdrSize does not take SDS but take a type?

This is a utility function. There are different cases where we need to get the header size when we still do not have full sds. immediate examples that pops to my head are  when we first create a new sds (_sdsnewlen) and when we need to adjust sds type (adjustTypeIfNeeded).

> // Why dont we have sdsHdrSizeByType

not sure I understand. doesn't sdsHdrSize do that?